### ISSUE TITLE
E3032 triggering on Lambda Layers despite mutually exclusive if statements

### ISSUE THREAD

**Issue by dothomson** [aws-cloudformation/cfn-lint#4458]
### CloudFormation Lint Version

1.48.1

### What operating system are you using?

Mac

### Describe the bug

Template linting started failing with: `E3032 expected maximum item count: 5, found: 6` 

The template has been the same for many months. Linting has only begun to fail recently. 

The linting was actually being run through `sam validate --lint` on our `ubuntu-24.04` github runner. 

The change which caused this to fail will have been one of the versions here: https://github.com/actions/runner-images/blob/main/images/ubuntu/Ubuntu2404-Readme.md 

---- 

Once able to reproduce the error locally it was due to the number of *Lambda Layer definitions* we had. 

The lambda layers are configured with `!If` statements. These if statements are mutually exclusive - so only one will be true at a time. 

```yaml
Layers:
  - !If 
    - Condition1
    - !Ref Layer1
    - !Ref AWS::NoValue
  - !If 
    - Condition2
    - !Ref Layer2
    - !Ref AWS::NoValue
```

etc. 

The conditions are based off a parameter which has a limited number of values. It is possible to prove they are mutually exclusive. 

I'm unsure if the changes behaviour is due to improved handling of the `Layers:` property - or there is a change in logic which has made `cfn-lint` unable to determine which conditions should be true. 

---- 

We have a valid work around which is to nest the conditions: 

```
- !If
  - Condition1
  - !Ref Layer1
  - !If
    - Condition2
    - !Ref Layer2
//   ...etc
```

However I am just flagging here in-case it helps anyone or indicates another issue. 

### Expected behavior

When calculating the number of Layers the maximum number of layers defined is determined by possible parameters and conditions. This would show that in-fact only two layers are ever defined in this example. 

(It is also possible that to correctly determine the maximum number of layers defined over all possible parameters + conditions would be computationally limiting. So no worries if that is the case!) 

### Reproduction template

```yaml
AWSTemplateFormatVersion: "2010-09-09"

Parameters:
  Environment:
    Type: String
    AllowedValues:
      - env1
      - env2

Conditions:
  Condition1: !Equals
    - !Ref Environment
    - env1

  Condition2: !Equals
    - !Ref Environment
    - env2

Transform: AWS::Serverless-2016-10-31

Resources:
  Function:
    Type: AWS::Serverless::Function
    Properties:
      FunctionName: function
      Handler: handler
      Layers:
        - !Ref LayerPersistent
        - !Ref LayerPersistent
        - !Ref LayerPersistent
        - !Ref LayerPersistent
        - !If
          - Condition1
          - !Ref Layer1
          - !Ref AWS::NoValue
        - !If
          - Condition2
          - !Ref Layer2
          - !Ref AWS::NoValue
      Runtime: nodejs24.x

  LayerPersistent:
    Type: AWS::Serverless::LayerVersion
    Properties:
      ContentUri: layer
      LayerName: !Sub ${AWS::StackName}-layer
      PublishLambdaVersion: true

  Layer1:
    Type: AWS::Serverless::LayerVersion
    Properties:
      ContentUri: layer1
      LayerName: !Sub ${AWS::StackName}-layer
      PublishLambdaVersion: true
    Condition: Condition1

  Layer2:
    Type: AWS::Serverless::LayerVersion
    Properties:
      ContentUri: layer2
      LayerName: !Sub ${AWS::StackName}-layer
      PublishLambdaVersion: true
    Condition: Condition2
```

**Comment by dothomson** [comment]
I've had a quick look through the release log and I believe this must be the source of my linting error! 

https://github.com/aws-cloudformation/cfn-lint/pull/4424 

I don't know the internals of the project and thus i don't know how feasible it is to perform this condition calculation on lambda layers. If it isn't then please feel free to close this bug report and my apologies for misattributing this improved functionality it as a bug 🙇

**Comment by kddejong** [comment]
Looking into this.
